### ISSUE TITLE
Unable to prepare context: path "." not found while using OneDrive Files On Demand

### ISSUE THREAD

**Issue by xposix** [docker/for-win#1290]
### Expected behavior
Build any image from any directory that contains a Dockerfile file.

### Actual behavior
I can't build an image from a folder using new "Filles On Demand" Onedrive feature with Windows 10 Fall Creators Update. I get the error message 'Unable to prepare context: path "." not found'
It doesn't work on any of my two computers with this feature enabled. All my Dockerfiles are set to "Always keep on this device" on both devices, but still doesn't work.

It works fine when I move the Dockerfiles out of Onedrive and when I redirect the Dockerfile to the build command, (example: docker build -t test - < Dockerfile)

### Information
  - Diagnostic ID from "Diagnose & Feedback" in the menu.
2E1C9248-2442-4E95-A494-A1A3CED026A7/2017-11-06_21-02-18
  - a reproducible case if this is a bug, Dockerfiles FTW
  - page URL if this is a docs issue or the name of a man page
  - host distribution and version (Windows version, build number, etc)
Windows 10 Fall Creators Update 17.09
Microsoft OneDrive 17.3.7074.1023

### Steps to reproduce the behavior
1. Install latest version of Windows
2. Install latest version of Onedrive
3. Enable the Files on Demand feature
4. Move Dockerfiles inside Onedrive
5. try any docker build inside any folder inside Onedrive.

**Comment by xposix** [comment]
The problem still happens when I disable "Files On Demand" option in OneDrive. I cannot build from a file that is being sync by OneDrive

**Comment by sampaioletti** [comment]
Not sure if it will help, but I was having the same problem, decided to create a drive->folder link because i've had some issues with the forced OneDrive folder (not sure if its the spaces or if its not a 'normal' folder) So In my case I went from d:/OneDrive - MyAccount/Code to p:/Code where p was the virtual drive letter I assigned the folder(I used http://www.ntwind.com/software/utilities/visual-subst.html, but you can also use the cmd https://technet.microsoft.com/en-us/library/bb491006.aspx) and the problem went away. It still syncs properly and i'm able to build within the OneDrive directory.

Just trying to share a potential workaround. Hope it helps.

**Comment by TDabasinskas** [comment]
Hello,

It seems that upgrading to the latest version released today (`17.12.0-ce-rc2-win41 (14746)`) the issue is gone. Can anyone else confirm that building images from OneDrive for Business paths is working with the latest version?

**Comment by LaurentLesle** [comment]
I still have the issue on my Windows 10 desktop regardless if I am on One Drive folder or not.

Default Isolation: hyperv
Kernel Version: 10.0 16299 (16299.15.amd64fre.rs3_release.170928-1534)
Operating System: Windows 10 Enterprise

The version of docker is:
Version 17.12.0-ce-win46 (15048)
Channel: stable
0ac7325

Do you know if it is possible to display verbose/debug logs during docker build?

**Comment by cveld** [comment]
@sampaioletti I tried out your proposed workaround to no avail. I still get the error. I.e.
unable to prepare context: unable to evaluate symlinks in context path: readlink Z:\Projects: The process cannot access the file because it is being used by another process.

Windows 10.0.16299
Docker 17.12.0-ce-win47 (15139)
Linux containers

**Comment by anishsunkara** [comment]
@sampaioletti Tried your workaround but having the same issue. 
"unable to prepare context: unable to evaluate symlinks in context path: readlink Z:\docker: The system cannot find the file specified."
Any help with this would be highly appreciated. Thank you, 

Docker version 18.01.0-ce, build 03596f51b1
Windows 10.

**Comment by willwhite1** [comment]
I'm facing a similar issue, everything I read says docker doesn't really support adding items from outside the build context. The workaround I have seen is to build from the directory you want to copy files and supply -f to the dockerfile rather than building .
I'm thinking to have a build.ps1 (and build.cmd wrapper for ease) and an ARG in the dockerfile which the ps1 can feed the $PSScriptRoot into enabling a combination of both external directory and items co-located with the dockerfile..... will let you know if I get anywhere.

**Comment by willwhite1** [comment]
Ok, I ended up doing the following... not sure if it helps or not.
I have a folder container my docker project - C:\projects\willsdocker
I have another folder containing some items I was to COPY/ADD to container - C:\something

In my docker project I created a build script - C:\projects\willsdocker\build.ps1
This script takes in mandatory param 'artifactPath' (so 'C:\something' in this example)
I copy 'this' to a new directory created in the artifactPath Parent - C:\dockertmp
then my docker build command works by specifying -f and .. for context - 
docker build -f "$($newDir)/Dockerfile" .. -t something

Obviously your Dockerfile COPY/ADD commands have to cater for relative references based on the above solution.

**Comment by Leonardo18** [comment]
Anyone fixed this? I'm getting this error when i run command docker build run to a dockerfile in C:, the Dockerfile ok, not have extension like .txt and is in folder that i make the command. Windows 10, Docker version latest

**Comment by vladvarna** [comment]
Tried with Docker Toolbox, and it throws the same error...

**Comment by willwhite1** [comment]
I'm not sure docker consider it to be an error, but rather an explicit design principle.... you can work around it by wrapping your docker build command into a build script which organizes the file system prior to docker build invocation.... that is what I ended up having to do.

**Comment by SirSkorpan** [comment]
I had the same problem and solved it similarly to what @willwhite1 did, where I created a temporary folder to which I pointed docker build for its context and the copied all the files docker build needed to this folder. It seems it's enough to create a new folder, even a sub folder in OneDrive, for this "locking" problem to go away, at least for a [smaller example project](https://github.com/SirSkorpan/docker-build-example). Maybe there could be other issues with larger projects in which case I'd recommend creating a temporary folder in the home directory.
